lib/encoding/yaml/parser.toit

```diff
@@ -0,0 +1,1571 @@
+
+
+class Parser_ extends PegParserBase_:
+  named-nodes := {:}
+  forbidden-detected/int? := null
+
+  constructor bytes:
+    super bytes
+
+  can-read num/int -> bool:
+    return super num
+           and (not forbidden-detected or
+                -(bytes-since-mark forbidden-detected) >= num)
+
+  match-hex digits/int -> int?:
+    with-rollback:
+      start := mark
+      failed := false
+      while digits-- > 0:
+        if not ns-hex-digit:
+          failed = true
+          break
+      if not failed: return int.parse --radix=16 (string-since start)
+    return null
+
+  apply-props props value/ValueNode_ -> ValueNode_:
+    tag/string? := null
+    if props:
+      tag = props[0]
+      anchor := props[1]
+      if anchor: named-nodes[anchor] = value
+    value.tag = tag
+    return value
+
+  detect-forbidden:
+    if forbidden-detected: return
+    mark := mark
+    if start-of-line and
+       (c-directives-end or c-document-end) and
+       (match-chars B-LINE-TERMINATORS_ or s-white or l-eof):
+      forbidden-detected = mark
+    rollback mark
+
+  find-leading-spaces-on-first-non-empty-line -> int:
+    start-mark := mark
+    start-of-line := mark
+    while true:
+      if s-white: continue
+      if b-break:
+        start-of-line = mark
+        continue
+      break
+    result := bytes-since-mark start-of-line
+    rollback start-mark
+
+    return result
+
+  // Overall structure
+  l-yaml-stream -> List:
+    k := repeat: l-document-prefix
+    documents := []
+    if document := l-any-document: documents.add document
+
+    repeat: l-yaml-stream-helper documents
+
+    if not l-eof:
+      throw "INVALID_YAML_DOCUMENT"
+    return documents.map: | node/ValueNode_ | node.resolve
+
+  l-yaml-stream-helper documents/List -> bool:
+    with-rollback:
+      if (repeat --at-least-one: l-document-suffix):
+        repeat: l-document-prefix
+        if document := l-any-document: documents.add document
+        return true
+
+    if c-byte-order-mark: return true
+    if l-comment: return true
+    if document := l-explicit-document:
+      documents.add document
+      return true
+
+    return false
+
+  l-any-document -> ValueNode_?:
+    with-rollback:
+      if document := l-directive-document: return document
+      if document := l-explicit-document: return document
+      if document := l-bare-document: return document
+    return null
+
+  l-directive-document -> ValueNode_?:
+    with-rollback:
+      repeat --at-least-one: l-directive
+      if res := l-explicit-document:
+        return res
+
+    return null
+
+  l-explicit-document -> ValueNode_?:
+    with-rollback:
+      if c-directives-end:
+        if res := l-bare-document: return res
+        if s-l-comments: empty-node_
+    return null
+
+  l-bare-document -> ValueNode_?:
+    with-rollback:
+      if res := s-l-plus-block-node -1 BLOCK-IN_: return res
+    return null
+
+  l-document-prefix -> bool:
+    c-byte-order-mark
```

Comment:
I would have expected a roll-back here, since we don't require a c-byte-order-mark.
So something like:
```
l-document-prefix -> bool:
  try-parse c-byte-order-mark
  repeat: l-comment
  return true
```

Reply:
Well, it is optional, and it self-rolls-back if not matching. There is an option method that is better to use for clarity, but it will impact performance. I added optional: c-byte-order-mark for clarity.